redaxo/src/core/pages/profile.php

```diff
@@ -104,23 +106,38 @@
 if (rex_post('upd_psw_button', 'bool')) {
     if (!$csrfToken->isValid()) {
         $error = rex_i18n::msg('csrf_token_invalid');
-    } elseif (!$userpsw || !$userpswNew1 || $userpswNew1 != $userpswNew2 || !rex_login::passwordVerify($userpsw, $user->getValue('password'))) {
+    } elseif (!$userpsw || !rex_login::passwordVerify($userpsw, $user->getValue('password'))) {
+        $error = rex_i18n::msg('user_psw_error');
+    } elseif ($passkey) {
+    } elseif (!$userpswNew1 || $userpswNew1 != $userpswNew2) {
         $error = rex_i18n::msg('user_psw_error');
     } elseif (true !== $msg = $passwordPolicy->check($userpswNew1, $userId)) {
         $error = $msg;
     } elseif ($passwordChangeRequired && $userpsw === $userpswNew1) {
         $error = rex_i18n::msg('password_not_changed');
-    } else {
-        $userpswNew1 = rex_login::passwordHash($userpswNew1);
+    }
 
+    if (!$error) {
         $updateuser = rex_sql::factory();
         $updateuser->setTable(rex::getTablePrefix() . 'user');
         $updateuser->setWhere(['id' => $userId]);
-        $updateuser->setValue('password', $userpswNew1);
+
+        if ($passkey) {
+            [$passkeyId, $passkeyPublicKey] = $webauthn->processCreate($passkey);
+            $updateuser->setValue('passkey_id', $passkeyId);
+            $updateuser->setValue('passkey_public_key', $passkeyPublicKey);
+            $updateuser->setValue('password', null);
```

Comment:
Aktuell kann ein Benutzer immer nur entweder ein Passwort oder einen Passkey haben. Es kann auch nur ein Passkey pro Benutzer hinterlegt sein.
Passt das?

Hatte das Gefühl, dass ich mir damit die Implementierung erleichtere.

Reply:
Ich weiß nicht, wie das in der Realität sein wird. Ob es Nutzer gibt, die wirklich mehrere Passkeys hinterlegen würden. Oder ob nicht jeder, der Passkeys nutzt, die auch auf seinen Geräten synchronisiert o.ä. Bzw. teils kann man glaube ich auch auf dem Rechner sich über Passkeys, die auf dem Mobilgerät sind, einloggen. 

Bei Apple werden die Keys ja gesynct zwischen den Geräten.

Reply:
Ich denke es wird viele leute geben mit iphone aber windows rechner (beispiel). Ich denke es muss gehen

Reply:
Ich denke es könnte sinn machen alle passkeys für den user zu listen, damit man einzelne löschen kann (analog user sessions)

Reply:
Das bedeutet aber auch, man muss optional das Passwort trotz hinterlegtem Passkey behalten können, oder?
Denn wenn auf iPhone den Passkey hinterlegt, muss man sich auf Windows dann ja erneut per Passwort einloggen können, um auch dort den Passkey hinterlegen zu können.

(Wobei man auf Windows auch die iCloud-Passwörter verwenden kann: https://support.apple.com/de-de/guide/icloud-windows/icw2babf5e03/icloud. Ich vermute, dass dort dann auch die Passkeys enthalten sind, weiß es aber nicht.)

Das hat dann Auswirkungen auf die Password-Policy. Also wenn man Passwort und Passkey hat, und die Passwort-Regel gilt, dass alle X Monate das Passwort ausgetauscht werden muss, muss man das auch tun, wenn man sich über den Passkey eingeloggt hat.

Reply:
Ich könnte mir vorstellen dem user die wahl zu lassen
- nur passwort
- nur passkey
- passwort und passkey

Und die passwort policy greift nur in den varianten wo weiterhin ein passwort im spiel ist.

Ggf. Könnte man auch überlegen ob sich die policy nur auf das passwort aber nicht den account auswirkt (d.h. Bei abgelaufenem passwort geht login via passkey weiterhin)